src/v/iceberg/transform_utils.h

```diff
@@ -13,12 +13,24 @@
 #include "iceberg/values.h"
 
 namespace iceberg {
+class partition_spec_field_error : std::exception {
+public:
+    explicit partition_spec_field_error(std::string msg) noexcept
+      : msg_(std::move(msg)) {}
 
+    const char* what() const noexcept final { return msg_.c_str(); }
+
+private:
+    std::string msg_;
+};
 // Transforms the given value to its appropriate Iceberg value based on the
 // input transform.
 //
 // TODO: this is only implemented for the hourly transform on timestamp values!
 // This will throw if used for anything else!
 value apply_transform(const value&, const transform&);
 
+// Returns true if the given transform can be applied to the given primitive
+std::optional<partition_spec_field_error>
```

Comment:
nit: could we use a `checked`, or maybe call this `check_transform_errors()` or something? It's a bit confusing for `!can_transform(...)` to be the case where we can transform